### ISSUE TITLE
[Discussion] Online injection of accessories

### ISSUE THREAD

**Issue by snowdd1** [nfarina/homebridge#237]
## TLDR/BLUF
I tried with injecting accessories after the server started and had published the bridge. I think that could be a cool way for platforms not to wait a given time for the accessories to be discovered, and also allows online addition of new devices, and in case of non-auto-discoverable devices, the possibility to add devices using a web interface of the running server.

## The long story:
A while back we discussed the matters of platform integration (#95), and found that devices discovered after [publish](https://github.com/nfarina/homebridge/blob/master/app.js#L140) are not shown as accessories of the bridge. I had a closer look into the publish and [Accessory.addBridgedAccessory()](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Accessory.js#L156) methods to find pros and cons of calling them on a running HAP server.

I found out that in theory it should be possible to dynamically add accessories by adding them to the bridge, however, the first attempts failed and the bridge showed only the accessories it had during pairing. As I knew that when restarting the server, new accessories immediately show up, I looked into the [Advertiser](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Advertiser.js) and how it worked.

Seeing there that there is a [configuration counter in the txtRecord](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Advertiser.js#L35) I made a small testing routine that adds accessories of the platforms delayed by 60 seconds each, and increasing the counter.

That however didn't work, until I stopped and restarted the bridge's advertiser after each configuration change, but then - **hey** - it worked like a charm.
Now I can see the new accessories pop up one every 60 seconds on my iPhone. 

Assuming having a nice web-based configuration interface I could now start the server and do my configuration by selecting plugins and such, just as mentioned in the [roadmap's last chapter](https://github.com/nfarina/homebridge/wiki/Roadmap#user-friendly-web-ui).

You can see the testing sample code [here](https://github.com/snowdd1/homebridge/commit/8358fb8778207df8692dc0084dfd2d1df971bb45)

The crucial lines are 
````javascript
  	  bridge.addBridgedAccessory(accessory);
  	  bridge._advertiser.stopAdvertising();
  	  bridge._advertiser.accessoryInfo.configVersion++;
  	  bridge._advertiser.accessoryInfo.save();
  	  bridge._advertiser.startAdvertising();
````
So far I had no incompliances with this setup. Draw back is that later added accessories seem sometimes to lose their room assignment (and other iOS DB related settings as use in scenes etc.) if an iOS App is currently listening to the events, so this is probably a reason **not to use this as a normal way** of instantiating known accessories but only for configuration purposes. If all apps are backgrounded on the iOS devices they *might not know* about the disappearance of an accessory and remember it right.

I will continue to explore that for my KNX platform, because configuration of that platform is very error prone in the config.json (see [example](https://github.com/nfarina/homebridge/blob/master/platforms/KNX-sample-config.json) ) and a nice web interface for adding accessories could help very much. 

@nfarina Do you plan to change the way platforms expose their accessories completely with the next API ? I still do not understand the provider-api branch sufficiently.

**Comment by KhaosT** [comment]
Glad to see you got this worked. Ideally the 60s delay should not be necessary but due to the lack of sending "update" support for the mdns package, you'll have to manually restart broadcasting after update configuration number. (https://github.com/agnat/node_mdns/issues/27)

For lose room, scene assignment, there might have some bugs in HAP-NodeJS on how we persist aid and iid, if those changed then HomeKit won't be able to recover room assignment for the specific accessory.

As I'm working on my own HomeKit app, I'm actually thinking about defining a bridge configuration service which allows user to associate accessory/platform directly from iOS instead of using a web interface.

**Comment by nfarina** [comment]
So here's the big problem with dynamic adding of accessories. Let's say you are happily running homebridge with 10 accessories that have been discovered, either at startup or later on. You've set up some scenes and triggers and stuff with HomeKit.

Now you restart homebridge. After startup homebridge must start publishing all the accessories it knows about - let's say after startup it knows about 5, then later the other 5 are added dynamically.

Meanwhile, as soon as the server comes up, iOS hears it and asks "what accessories do you have?" (`/accessories`). If it only hears about 5 accessories, then it will simply delete the other 5 "old" ones, and remove them from all triggers and scenes.

Later on, those 5 accessories are dynamically discovered again and added - but even if they retain the same AID/IIDs, I'm not sure what HomeKit will do. I'm guessing it's too late - it's already deleted all references to them.

If we do dynamic discovery, we'll have to figure out some kind of "caching" of all previously discovered accessories - along with some manual removal mechanism for the user to prune old accessories that really are gone (because say they got rid of them).

**Comment by snowdd1** [comment]
@KhaosT : the 60 seconds was just for testing purposes to simulate a late discovery or a later enablement of a device, to give me amble time to watch the log.

The experienced vanishing of accessories if the bridge starts up is a bit flaky. I had it the first time I retried with the new bridge configuration, but after that the accessory stayed in the room even when the iOS app was in foreground. It might just be a user error on my behalf, I might not have hit the back-button in eve at the right time, and that's apparently the point of time at which eve is persisting the configuration; there are no OK or Save buttons in ios style.

I will do several test on that to get a more confident opinion myself.

It might be a good idea to look in the persistence layer and return all devices as "unreachable" which is supposed a standard bridge behavior - a bridge must assume that a bridged accessory can be out of battery or off network coverage then and when. And then, I agree with Nick on that, have an extra mechanism to get rid of old entries e.g. Via web interface or whatsoever. 

Somewhere else I thought of a `removeBridgedAccessory()` method to handle that.

**Comment by snowdd1** [comment]
Losing the assignments is reliably after some time lag between bridge starting and advertising late accessories. So it is not as simple as in my little demo. :disappointed: It seems that if it is within some few seconds HomeKit remembers the assignments, and after that - boom - gone. Like a short-term memory loss (as in "50 first dates" years ago). 
The bridge has to report all formerly advertised accessories as unreachable (I think it's "s#":0 in [txtRecord](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Advertiser.js#L36), will check that.

**Comment by nfarina** [comment]
Is there a way to specify an accessory as "unreachable" when behind a bridge? This would have to be part of `/accessories` as `txtRecord` doesn't apply here (mdns is only for the bridge itself)

**Comment by snowdd1** [comment]
I think I heard of something like that. Probably in one of the session for homekit that were available as video. I cannot find any references in a wild-ass-Google-search.
But I think it has to have something like that. HomeKit can quite easily deal with accessories out of reach (which it should as a mobile device tool) but the same might happen to bridged accessories as for example simply their battery might die. 
I keep on searching. I already noticed that the txtRecord and the AccessoryInfo persistence is for the bridge only, and we persist only the identifier cache (which probably would be to few info to keep the memory of homekit alive)

**Comment by snowdd1** [comment]
I didn't find any traces of that in the Internet after an hour of searching. Maybe I just dreamed of it.
The probably easiest way to cope with that would be to expose all accessories not as bridged but single accessories, multiplying the network traffic but making use of HomeKits inherent ability to accept that devices might be out of reach.
It's getting late here, I will think it over tomorrow.

**Comment by KhaosT** [comment]
Yeah, currently there is no way for a bridge to inform HomeKit that a bridged accessory is not reachable. It shouldn't hurt if we just persist the accessory that has been added to the bridge and reload all those accessories from persist storage when the bridge rebooted.

**Comment by justme-1968** [comment]
such an app could also allow to push group and room names from the HAP/homebridge side to homekit and changes to names on the homekit side back to homebridge.

this would be great for bridges that bridge a complete home automation system that already has rooms and groups defined. 

so adding a bridge to homekit with this app could query homebridge for the rooms that the accessories are supposed to be in and use these as a default on the homekit side instead of entering them all manually or using a default room like eve does.

if things like rooms,groups and names would be pushed back they could be saved on the homebridge side and reused. this would ease the pain of loosing the connection a little bit. 

the dynamically appearing and disappearing devices behind a bridge was visible in the wwdc video. i don't think unreachable on the accessory level is handled by mdns as there is only one entry for the whole bridge.

by the way: loosing connection... i just had a power outage. after reboot of the homebridge server homekit had lost all devices. did not even start a connection. but unlike the other times i have lost the devices this time a delete device on the homkit side did work. i could see a nde connection and an unpair.
also loosing devices is probably related to the number of devices used. it is nearly perfectly stable if i use only a few devices but very unstable if i add more devices. maybe there is indeed still something wrong with the persist layer.

**Comment by snowdd1** [comment]
I knew I saw it somewhere. In a discussion in slack @KhaosT mentioned the Service.BridgingState and Characteristic.Reachable that I meant. However, I still have no idea how to use that.

**Comment by cflurin** [comment]
@snowdd1: I tried this service with the homebridge-punt simulator. EVE can detect the accessory but  doesn't support it.

config-punt.json definition:
```
 {
   "name": "fack_bridge",
   "service": "BridgingState"
 },
```

**Comment by snowdd1** [comment]
@cflurin Did you try that service on an existing bridged accessory or just stand alone? I would think it's an additional service 
https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L1374
 for a bridged accessory that can be reachable or unreachable.
https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L874
Maybe I find time to check that myself.

**Comment by cflurin** [comment]
@snowdd1 I defined a new accessory but I think it doesn't matter. The characteristic Reachable can be true or false that's all. I tested it with HCatalog. If the value is true, you can see the name otherwise it's grayed out.
I suppose the purpose is to set Reachable in your plugin if the bridge is not sending data. E.g. the plugin for the Philips hue bridge could set Reachable false if the connection to bridge is down. Anyway the app should do something with this value.

**Comment by snowdd1** [comment]
@cflurin You're right. at least the app should not try to send data to an unreachable device. I am wondering if homekit itself would try to do so, or if the status has some magic in homekit itself. 
It should be the solution to @KhaosT 's https://github.com/nfarina/homebridge/issues/237#issuecomment-146325763 - inform homekit that a bridged device is currently unavailable.
This discussion thread was dealing with online injection of bridged accessories, and during my test I found out that (temporary) removal of bridged accessories makes homekit delete them and therefore wreck scenes and such with theses devices in it.

To avoid that, the plugin that is lately binding (injecting) accessories should assume that all devices (discovered last run time) are still there but did not answer yet, and therefore return all of them as "currently unavailable" (unreachable, but existent) at startup. When the devices are discovered, known devices are set to "available" (reachable), and new devices are injected into HAP and the bridges advertiser restarted.

**Comment by pdlove** [comment]
When I tested this with the EVE app, it ignored the Reachable flag. In the wink plugin, we simply return an error if wink has marked the device as disconnected.
I've been thinking about this alot because I want devices to dynamically add/remove as they are added or removed with Wink. I've got some changes to hap-nodejs I'm refining to get HomeKit to immediately repoll accessories by disconnecting the connections after re-advertising. Just re-advertising works, but the device list only gets updates when a change is submitted.

I've also been thinking about how this relates to the WeMo issue. The core problem with it is it doesn't know what devices it broadcasted previously when it is first started. I didn't have this issue with the wink plugin and won't have it with the smartthings platform because they are tracked by the service and the service returns if it is connected or not.

I think @snowdd1 is completely correct with the solution to that. when devices are sent to advertise, they should also be saved in persistent cache and that list used as a basis next time it is launched. There would need to be some kind of device removal timeout managed by the plugin.

**Comment by rudders** [comment]
We are working on a persist version of platform-wemo. @Devbobo is the lead on it. He's away for another week and when he's back we'll tidy it up and release. Issues around usability are the only remaining concerns for me, he's got the code nailed (nearly)

**Comment by rudders** [comment]
IOS 9.3 solves much of this I believe too.

**Comment by KhaosT** [comment]
Alright, spent sometime experience the possibility to make this reliable. Now in the branch (https://github.com/KhaosT/HAP-NodeJS/tree/bridgeUpdate), it's possible to add/remove accessory from a bridge dynamically and it's possible to update reachability of individual accessories.

From homebridge's aspect, I'd propose the solution as following:
Add a new version of API for plugin which no longer allows plugin to directly add/remove accessory.
Instead, new APIs like ``registerAccessory`` and ``unregisterAccessory`` will be provided to plugin.
homebridge will now caches all registered accessories (all services, characteristics) and when homebridge launches, it will publishes all accessories it saved. Plugin will be able to get back specific accessory from homebridge by ``UUID``.

The new launch procedural will be come

homebridge boot -> load all installed plugin -> load saved accessories -> based on accessory, invoke associated plugin to handle accessory -> set all saved accessories that plugin can't handle to not reachable -> publish HAP server

In this way, user should no longer need to worry that the temporary disappear of accessory from network will remove accessory related settings disappear from HomeKit. And now it's possible for plugin to dynamically add/remove accessories explicitly.

A bridge configuration service will be defined and can be used to register/unregister accessory via HomeKit app.

**Comment by nfarina** [comment]
Crazy - so HomeKit Catalog for instance has added UI for adding and removing accessories connected to bridges?

**Comment by KhaosT** [comment]
@nfarina sorry for confusion, the bridge configuration service I talked at here is different from the official one in 9.3. The official one is too limited for our case.

**Comment by pdlove** [comment]
So, just to clarify, the new API would have AddAccessory, GetAccessories and RemoveAccessory?
When my platform starts up, I can get the cached accessories, determine if there are any new ones and add them, then determine if any have been removed from the platform and, if so, remove them. Would the Accessory store also retain all of my extra data on the Accessory or would it only store the UUID and HomeKit-related items?
I think it sounds perfect.

**Comment by KhaosT** [comment]
@pdlove the proposal is still at it's early stage but those 3 APIs should be enough (if we want to let homebridge handle the setup without user editing config.json then there will be 2 more I guess) for homebridge to maintain it's current functionality while solving this issue.

About ``unregisterAccessory``, we still need to decide whether plugin itself can invoke that or it has to by done by user (via web/app). It can be really hard for plugin to figure out whether the accessory is just temporarily disappeared or the accessory is actually gone forever. Instead of encourage plugin to simply delete all accessories it didn't see (by giving plugin the access to remove accessory), maybe this should only be done by user.

Ideally homebridge should just take a snapshot of all HomeKit related objects like accessories, services and characteristics that's enough to reconstruct the accessory tree but a persistent context on the each object may also be saved if plugin developer really need that...

**Comment by pdlove** [comment]
I think a developer should be good with just the UUID, Services and Characteristics. The main reason I'd like the platform to have the ability to remove the device is because with the ones I'm working with (Wink, Smartthings, Ecobee, etc) report to me all devices, even ones that aren't currently online, so I can know for sure that a device is gone and isn't going to come back in a way that I'll be able to reconnect with the persistent devices. I'm all for removing user steps when possible. The Web Interface would be great as a plugin. I've toyed with making a web site that would use these home bridge plugins as the source so I can give my wife, who is an Android user, and my kids who only have Windows Tablets the ability to interact with the completely integrated home automation solution.
Thank you guys for the hard work so far!

**Comment by snowdd1** [comment]
Very good, I saw that https://github.com/KhaosT/HAP-NodeJS/pull/195 has implemented the reachability characteristic and an updated txt record updater, so the work around with stopping and restarting the advertiser isn't necessary any more - that meaning that homebridge doesn't have to expose it's advertiser object to the plugins.

**Comment by snowdd1** [comment]
Even cooler, I just found the time to read the new accessory.js implementation. You have been very busy while I spent my time at my ill son's hospital bed. As far as I can see everything this discussion was about has been solved. 
Closing this now!
I am very keen to test all that with my plugin when I find time to get to that again!
